### DISCUSSION TITLE
How can we make chat app example alive for all time?

### DISCUSSION THREAD

**Discussion by pawanl1x** [libp2p/rust-libp2p#4553] [Q&A]
The peers in the chat example are getting removed from the network after being inactive for some duration. Is there any way to make it alive for all time?

I've tried setting mDNS's TTL as Max but the new() of mDNS uses the default value of TTL in the discoverd_nodes config item.

**Comment by thomaseizinger** [comment]
You can increase the idle connection timeout on the `SwarmBuilder`. See here in the ping-example: https://github.com/libp2p/rust-libp2p/blob/636966697ea89c333f15148220d88a5f46e91f09/examples/ping/src/main.rs#L47

See https://github.com/libp2p/rust-libp2p/issues/4306 for the tracking issue around this topic.

**Reply by pawanl1x** [reply]
Thanks for the response @thomaseizinger.

I checked this one but when I'm setting the max value, the program is panicking with the overflow error.

Is there anything related to before, I can see initially we had something related to keep_alive()?

Or what max value I can set in order to keep my application alive even after not getting any event for long time?

**Reply by thomaseizinger** [reply]
> I checked this one but when I'm setting the max value, the program is panicking with the overflow error.

This sounds like a bug. Can you open an issue for this?

> Or what max value I can set in order to keep my application alive even after not getting any event for long time?

How long do you want your connections to be alive when they are idle? A day? A week? `u64::MAX` is 584942 million years. We shouldn't panic when you set that but I am sure there is a smaller value that will work for you :)

**Reply by pawanl1x** [reply]
Sure, thanks.

Let me open a bug for this.

Yes, I need the maximum value which I can configure.

FYI: I tried to set it for a week and got this error after :
thread 'main' panicked at 'record types do not match, NSEC <> Some(Unknown(47))', /.cargo/registry/src/index.crates.io-6f17d22bba15001f/trust-dns-proto-0.23.0/src/rr/resource.rs:537:9

**Reply by thomaseizinger** [reply]
That is an unrelated, known issue: https://github.com/libp2p/rust-libp2p/issues/4507

**Reply by pawanl1x** [reply]
Do we have any other approach to make our connection alive?

I mean related to gossipsub::publish so that every node makes itself active in the network?

**Reply by thomaseizinger** [reply]
> Do we have any other approach to make our connection alive?

Use them! As long as there is activity on the connections, they won't be closed automatically. Can you share your usecase of why you need connections to be kept alive beyond that? What would be the issue with re-establishing them if you need to send a message to a specific peer?

> I mean related to gossipsub::publish so that every node makes itself active in the network?

I don't understand what you mean here, can you elaborate?

---

The overflow panic you are experiencing shouldn't be too hard to fix if you are interested in contributing. I comment on the issue you opened with an idea on how to!

**Reply by pawanl1x** [reply]
I mean, do we have any other option to make peers alive apart from idle_connection_timeout?
BTW, I saw your comment, we just need to wait for the release now.

Regarding the overflow issue, sure, I will take a look.

**Reply by thomaseizinger** [reply]
> I mean, do we have any other option to make peers alive apart from idle_connection_timeout?

We don't have another (recommended) mechanism to keep **idle** connections alive. There is `KeepAlive::Until` but we are fading that out: https://github.com/libp2p/rust-libp2p/issues/3844

Can you share your usecase why you would like to keep **idle** connections alive? In other words, connections that have no messages exchanged?

**Reply by pawanl1x** [reply]
My use-case is just gossip data into the network, but the issue right now is after 6 minutes peers are getting dropped. I just don't want to miss any data even if I get data after 1 hour or more than that.

**Reply by thomaseizinger** [reply]
Is this within a trusted network? i.e. are your peers definitely not malicious? Gossipsub doesn't guarantee that you are not going to miss any data. If a node crashes, it will miss all messages that are being gossiped while it reboots. If every node MUST witness all messages, you'd probably need some kind of "sync" protocol anyway that you can use to retrieve historic messages.

Also, how big is your network and how strong are your machines? Depending on the size of the network, you might be able to maintain full connectivity and don't need to gossip data but can just request it directly from other peers.
